File: display.py
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import geopandas as gpd  # For maps
import json


data = [
    [350.0, 35000.0, 5.0, [0, 0], 0.1, 18.0, 0.5],
    {
        1: [200.0, 6.84931506849315e-05, 200],
        2: [160.0, 0.000273972602739726, 200],
        3: [140.0, 0.0008219178082191781, 200],
    },
    {1: [80, 0]},
    {1: [40.0, 0.0], 2: [35.0, 0.0]},
    {
        1: [0.0, 5.479452054794521e-05, 200, 0.015],
        2: [0.0, 5.479452054794521e-05, 300, 0.045],
        3: [0.0, 0.000273972602739726, 200, 0.01],
        4: [0.0, 0.000273972602739726, 300, 0.03],
        5: [0.0, 0.0008219178082191781, 200, 0.0085],
        6: [0.0, 0.0008219178082191781, 300, 0.0255],
    },
    {
        1: [0.0, 100, 0.01],
        2: [0.0, 200, 0.02],
        3: [0.0, 100, 0.009000000000000001],
        4: [0.0, 200, 0.018000000000000002],
        5: [0.0, 100, 0.0085],
        6: [0.0, 200, 0.017],
    },
]


def display_data(data):
    # wind turbines = red
    # substation = blue
    # cable = black
    # land station = green

    V_t = data[2]
    V_s = data[3]

    # Create a figure and axis
    fig, ax = plt.subplots()

    r = 0.1

    # Draw a red filled circle for each wind turbine
    for v in V_t.values():
        circle = plt.Circle(v, r, fill=True, color="red")
        ax.add_patch(circle)

    # Draw a blue circle for each substation
    for v in V_s.values():
        circle = plt.Circle(v, r, fill=True, color="blue")
        ax.add_patch(circle)

    # Draw a green circle for the land station (in 0,0)
    circle = plt.Circle([0, 0], r, fill=True, color="green")
    ax.add_patch(circle)

    # Adapt the x axis
    plt.xlim(-5, 85)
    plt.ylim(-10, 10)

    # Display the plot
    plt.show()


# draw_data(data)


def display_solution(x, y, z, I):
    V_t = I[2]
    V_s = I[3]
    S = I[1]

    # Create a figure and axis
    fig, ax = plt.subplots()

    r = 0.1

    # Draw a red filled circle for each wind turbine
    for t in V_t.values():
        circle = plt.Circle(t, r, fill=True, color="red")
        ax.add_patch(circle)

    # Draw a blue circle for each substation
    for s in V_s:
        # s'il existe une sous-station ie il existe j tel que x[(s,j)]==1
        for j in S:
            if x[(s, j)] == 1:
                # Draw a line from the substation to the land
                plt.plot([V_s[s][0], 0], [V_s[s][1], 0], color="green")
                # Draw a line from the substation s to the turbines t si z[(s,t)]==1
                for t in V_t:
                    if z[(s, t)] == 1:
                        plt.plot(
                            [V_s[s][0], V_t[t][0]],
                            [V_s[s][1], V_t[t][1]],
                            color="red",
                        )
            circle = plt.Circle(V_s[s], r, fill=True, color="blue")
            ax.add_patch(circle)
    for s in V_s:
        circle = plt.Circle(V_s[s], r, fill=True, color="blue")
        ax.add_patch(circle)

    # Draw a green circle for the land station (in 0,0)
    circle = plt.Circle([0, 0], r, fill=True, color="green")
    ax.add_patch(circle)

    # Adapt the x axis
    plt.xlim(-5, 85)
    plt.ylim(-10, 10)

    # Display the plot
    plt.show()


def display_type_ss(I):
    S = I[1]
    fig, ax = plt.subplots()

    costs = [s[0] for s in S.values()]
    failure = [s[1] for s in S.values()]
    rating = [s[2] for s in S.values()]
    id = [s for s in S.keys()]

    # Sur chaque point, ajouter une étiquette avec son identifiant
    for i, txt in enumerate(id):
        ax.annotate(txt, (costs[i], failure[i]))
    plt.scatter(costs, failure, c=rating, cmap="viridis")
    plt.xlabel("Cost")
    plt.ylabel("Failure probability")

    # add legend for rating
    plt.colorbar()

    plt.show()


def display_type_cable_land_ss(I):
    pass


# def load_sol(file_path):
#     with open(file_path, "r") as json_file:
#         data = json.load(json_file)

#     # Initialiser les structures de données
#     x = {}
#     y = [[], []]
#     z = {}

#     # Remplir les structures de données à partir du contenu du fichier JSON
#     for substation in data.get("substations", []):
#         substation_id = substation["id"]
#         substation_type = substation["substation_type"]
#         x[(substation_id, substation_type)] = 1

#     for cable in data.get("substation_substation_cables", []):
#         substation_id = cable["substation_id"]
#         other_substation_id = cable["other_substation_id"]
#         cable_type = cable["cable_type"]
#         y[1].append((substation_id, other_substation_id, 0, cable_type))

#     for turbine in data.get("turbines", []):
#         substation_id = turbine["substation_id"]
#         turbine_id = turbine["id"]
#         z[(substation_id, turbine_id)] = 1

#     return x, y, z

# # Exemple d'utilisation
# file_path = "solutions/votre_solution.json"
# x, y, z = load_sol(file_path)

File: display.jl
include("utils.jl")
include("instance.jl")
include("solution.jl")
include("parsing.jl")
include("eval.jl")
include("sixtine_sol.jl")

using Plots
gr()

ms2 = 2


function display_instance(instance::Instance)
    V_t = instance.wind_turbines
    V_s = instance.substation_locations

    # Create a plot
    plot()

    r = 0.02

    # Draw a red filled circle for each wind turbine
    for (id, turbine) in pairs(V_t)
        plot!([turbine.x], [turbine.y], st=:scatter, ms=ms2, m=:circle, color="red")
        #annotate!(turbine.x, turbine.y, text(string(id), :left, 4, :red))
    end

    # Draw a blue circle for each substation
    for (id, substation) in pairs(V_s)
        plot!([substation.x], [substation.y], st=:scatter, ms=ms2 * 4, m=:circle, color="blue")
        annotate!(substation.x, substation.y + 0.3, text(string(id), :bottom, 8, :blue))
    end

    # Draw a green circle for the land station (in 0,0)
    land = instance.land
    plot!([land.x], [land.y], st=:scatter, ms=ms2 * 5, m=:circle, color="green", legend=false)

    # Set axis limits
    # plot!([-5, 85], [-10, 10], legend=false)

    # Display the plot
    display(plot!())
end

function display_solution(solution::Solution, instance::Instance)
    V_t = instance.wind_turbines
    V_s = instance.substation_locations

    plot()

    for (id, substation) in pairs(V_s)
        plot!([substation.x], [substation.y], st=:scatter, ms=ms2, m=:circle, color="blue")
    end

    # Draw a blue circle for each built substation
    for substation in values(solution.substations)
        ss = V_s[substation.id]
        plot!([ss.x], [ss.y], st=:scatter, ms=ms2 * 4, m=:circle, color="blue")
        annotate!(ss.x - ms2, ss.y, text(substation.substation_type, :bottom, 8, :blue))
        plot!([ss.x, 0], [ss.y, 0], color="green")
        annotate!((ss.x + 0) / 2, (ss.y + 0) / 2, text(substation.land_cable_type, :top, 8, :green))
    end

    # Draw a green circle for the land station (in 0,0)
    land = instance.land
    plot!([land.x], [land.y], st=:scatter, ms=ms2 * 5, m=:circle, color="green", legend=false)

    # Draw the links between substations
    for i in 1:nb_station_locations(instance)
        for j in (i+1):nb_station_locations(instance)
            if solution.inter_station_cables[i, j] > 0
                plot!([V_s[i].x, V_s[j].x], [V_s[i].y, V_s[j].y], color="black")
                annotate!((V_s[i].x + V_s[j].x) / 2, (V_s[i].y + V_s[j].y) / 2, text(solution.inter_station_cables[i, j], :bottom, 8, :black))
            end
        end
    end

    # Draw a red filled circle for each wind turbine
    for id_t in 1:nb_turbines(instance)
        id_s = solution.turbine_links[id_t]
        plot!([V_s[id_s].x, V_t[id_t].x], [V_s[id_s].y, V_t[id_t].y], color="red")
        # annotate!((V_s[id_s].x + V_t[id_t].x) / 2, (V_s[id_s].y + V_t[id_t].y) / 2, text(id_t, :top, 4, :red))
        plot!([V_t[id_t].x], [V_t[id_t].y], st=:scatter, ms=ms2, m=:circle, color="red", legend=false)
        annotate!((V_s[id_s].x + V_t[id_t].x) / 2, (V_s[id_s].y + V_t[id_t].y) / 2, text(id_t, :right, 4, :red))
    end

    # Set axis limits
    # plot!([-5, 85], [-10, 10], legend=false)

    # Display the plot
    title!("Cost: " * string(cost(solution, instance)))
    display(plot!())
end


function display_type_ss(I::Instance)
    plot()

    for s in values(I.substation_types)
        plot!([s.cost], [s.probability_of_failure], st=:scatter, ms=ms2, m=:circle, color="blue")
        annotate!(s.cost, s.probability_of_failure, text(s.id, :bottom, 8, :blue))
    end

    xlabel!("Cost")
    ylabel!("Failure probability")
    title!("Type of substations possible")

    display(plot!())
end


function display_type_cable_land_ss(I::Instance)
    plot()

    # tot_costs = [q.variable_cost + q.fixed_cost for q in values(Q_0)]
    # failure = [q.probability_of_failure for q in values(Q_0)]
    # rating = [q.rating for q in values(Q_0)]

    for q in values(I.land_substation_cable_types)
        plot!([q.variable_cost + q.fixed_cost], [q.probability_of_failure], st=:scatter, ms=ms2, m=:circle, color="green")
        annotate!(q.variable_cost + q.fixed_cost, q.probability_of_failure, text(q.id, :bottom, 8))
    end

    # scatter!(tot_costs, failure, c=rating, cmap="viridis")
    xlabel!("Total cost")
    ylabel!("Failure probability")
    title!("Type of cables possible (land -> substation))")

    display(plot!())
end




# ------------- Exemple d'utilisation -------------


txt = "KIRO-huge"
I = read_instance("instances/" * txt * ".json")
sol::Solution = resolution_sixtine(I)
write_solution(sol, "solutions/" * txt * "2.json")
# solution = read_solution("solutions/" * txt * ".json", I)
solution2 = read_solution("solutions/" * txt * "2.json", I)


# display_instance(I)
# display_solution(solution, I)
display_solution(solution2, I)

# print(cost(solution, I))
# print(cost(solution2, I))

# display_type_ss(I)
# display_type_cable_land_ss(I)

